### ISSUE TITLE
Great Circle Lines Across Dateline

### ISSUE THREAD

**Issue by JackHenry99** [dceejay/RedMap#216]
Hi

The issue with drawing great circle lines across the dateline has 'changed' slightly. It draws the line, but the pin gets left behind on another part of the map. See images. Origin point near Adelaide South Australia. Target mid USA. This is with the latest version of World Map

<img width="320" alt="Dateline-1" src="https://user-images.githubusercontent.com/13172326/204122628-950b3954-7935-4b79-836a-b868ec735581.png">
<img width="320" alt="Dateline-2" src="https://user-images.githubusercontent.com/13172326/204122629-f4f3fcad-d05b-4112-b464-5d8c796eccbc.png">

**Comment by dceejay** [comment]
yeah leaflet doesn't handle it well... need some inspiration to fix it. May take a while unless someone can spot/suggest anything clever

**Comment by mrdvt92** [comment]
USGS has a pair of programs called forward and inverse that calculate the heading between two points and then the terminus for a point and heading vector.  We could use those algorithms to find the point on the transition and solve the problem.  I ported them to Perl as while back.  https://metacpan.org/pod/Geo::Forward and https://metacpan.org/pod/Geo::Inverse

**Comment by JackHenry99** [comment]
Wasn't v2.27.3 supposed to fix the issue??

**Comment by dceejay** [comment]
It "improved" the issue - before it was totally broken and created loads of horizontal lines... 
Actually looking at that image then yeah... if you are zoomed out that far leaflet only plots one marker... so not sure there is anything to do about that... it doesn't handle  (say) 90 and 450 as two different points 360 degrees apart... I don't think you get to choose which version of America you get to put the maker on...  You have to zoom in so you only have one world...

**Comment by JackHenry99** [comment]
> Are you able to share the necessary part of the flow that creates that issue ?

deceejay

I THINK this is the part you need.  I'm not sure how to export just a sub-flow, so this is the dashboard tab from the entire thing. The map is in the top block of flows. 

[Dashboard.json.zip](https://github.com/dceejay/RedMap/files/10107970/Dashboard.json.zip)

Regards
John VK5JAK

**Comment by dceejay** [comment]
Sorry - no I just mean a small flow that had the necessary end points to show the problem...  Ie just injected the necessary draw command . Your flow talks to some database that I don't have (nor want :-) )  and clicking the inject buttons doesn't create any useful markers. Can you maybe add a debug just before the map (maybe the msg one there already ? ) - send it some data that draws the bad great circle and copy the payload output from that debug here as text ?

**Comment by JackHenry99** [comment]
dceejay

This is the debug from the INPUT side of the world map. There is no debug on the output. The logged callsign is KK6USY and on the map image below, is the Western of the two flags on the USA. I've circled on the flow the node the debug is from.

Regards
John VK5JAK (VK5Y)

{"payload":{"name":"KK6USY","lat":"38.00196","lon":"-122.28970","ttl":"1200","op":"VK5Y","band":"10M","mode":"SSB","isrunqso":"0","exchange":"CA","section":"CA","distance_km":13063,"distance_miles":8117,"iconColor":"#FF8C00","layer":"10M","label":"10M","ismultiplier":0,"greatcircle":[[-35.05,138.08],[38.001961,-122.289703]]},"fromip":"10.1.1.5:51673","ip":"10.1.1.5","port":51673,"_msgid":"c70fcaaffe082120","qsotype":"qso","_event":"node:677d263669792c2d","original":{"app":["N1MM"],"contestname":["ARRL10M"],"contestnr":["1"],"timestamp":["2022-11-30 02:28:13"],"mycall":["VK5Y"],"band":["28"],"rxfreq":["2850060"],"txfreq":["2850060"],"operator":["VK5Y"],"mode":["USB"],"call":["KK6USY"],"countryprefix":["K"],"wpxprefix":["KK6"],"stationprefix":["VK5Y"],"continent":["NA"],"snt":["59"],"sntnr":["25"],"rcv":["59"],"rcvnr":["0"],"gridsquare":[""],"exchange1":["CA"],"section":["CA"],"comment":[""],"qth":[""],"name":[""],"power":[""],"misctext":[""],"zone":["3"],"prec":[""],"ck":["0"],"ismultiplier1":["0"],"ismultiplier2":["0"],"ismultiplier3":["0"],"points":["0"],"radionr":["1"],"run1run2":["1"],"RoverLocation":[""],"RadioInterfaced":["0"],"NetworkedCompNr":["0"],"IsOriginal":["True"],"NetBiosName":["WINDOWS-MACHINE"],"IsRunQSO":["0"],"StationName":["WINDOWS-MACHINE"],"ID":["4e3be4493fb745748f4868eb05399a99"],"IsClaimedQso":["1"],"oldtimestamp":["2022-11-30 02:28:13"],"oldcall":["KK6USY"]},"qrzkey":"9d0b13e696977eb71ff9420b5e375e34","statusCode":200,"headers":{"date":"Wed, 30 Nov 2022 02:28:12 GMT","server":"Apache/2.4.29 (Ubuntu)","expires":"Tue, 29 Nov 2022 02:28:12 GMT","vary":"Accept-Encoding","access-control-allow-origin":"*","connection":"close","transfer-encoding":"chunked","content-type":"text/xml; charset=utf-8","x-node-red-request-node":"7fabbb1a"},"responseUrl":"https://xmldata.qrz.com/xml/current/?s=9d0b13e696977eb71ff9420b5e375e34;callsign=KK6USY","redirectList":[],"retry":0,"QRZDatabase":{"$":{"version":"1.36","xmlns":"http://xmldata.qrz.com"},"Callsign":[{"call":["KK6USY"],"dxcc":["291"],"fname":["Charles M"],"name":["Thompson"],"addr1":["1170 Willow St, PO BOX 1072"],"addr2":["Pinole"],"state":["CA"],"zip":["94564"],"country":["United States"],"lat":["38.001961"],"lon":["-122.289703"],"grid":["CM88ua"],"county":["Contra Costa"],"ccode":["271"],"fips":["06013"],"land":["United States"],"efdate":["2015-10-27"],"expdate":["2025-06-16"],"class":["G"],"codes":["HAI"],"email":["gliderunner@gmail.com"],"u_views":["10044"],"bio":["1598"],"biodate":["2020-09-03 03:41:01"],"image":["https://cdn-xml.qrz.com/y/kk6usy/IMG_2421.JPG"],"imageinfo":["426:640:140576"],"moddate":["2020-09-04 03:07:19"],"MSA":["5775"],"AreaCode":["510"],"TimeZone":["Pacific"],"GMTOffset":["-8"],"DST":["Y"],"geoloc":["geocode"],"name_fmt":["Charles M Thompson"]}],"Session":[{"Key":["9d0b13e696977eb71ff9420b5e375e34"],"Count":["8499"],"SubExp":["Fri Apr 21 07:11:43 2023"],"GMTime":["Wed Nov 30 02:28:12 2022"],"Remark":["cpu: 0.021s"]}]}}




<img width="1081" alt="Screen Shot 2022-11-30 at 12 59 57" src="https://user-images.githubusercontent.com/13172326/204693244-b9a2dd50-0007-4940-992e-60234b04418c.png">


![Debug-highlight](https://user-images.githubusercontent.com/13172326/204693612-09ba95a0-043f-4dc0-a82e-db6e66c54415.jpg)

**Comment by dceejay** [comment]
Thanks - much better. Pushed version 2.31.2 with a fix
<img width="1147" alt="image" src="https://user-images.githubusercontent.com/5375409/204847179-2d2be280-263d-4d38-a310-146f22d4f9ff.png">

**Comment by JackHenry99** [comment]
What was the issue? Also, the image you posted looks to be the same issue. If Australia is centred on the map, does the line extend directly to USA? I can't test it for a couple of hours.
John
